external-crates/move/crates/move-vm-runtime/src/interpreter/eval.rs

```diff
@@ -0,0 +1,1156 @@
+        Vector, VectorRef,
+    },
+};
+use smallvec::SmallVec;
+
+use std::collections::VecDeque;
+
+const DEBUG_STEP_PRINT: bool = false;
+
+#[derive(PartialEq, Eq)]
+enum StepStatus {
+    Running,
+    Done,
+}
+
+struct RunContext<'loader, 'native, 'native_lifetimes> {
+    loader: &'loader Loader,
+    link_context: AccountAddress,
+    defn_resolver: ModuleDefinitionResolver<'loader>,
+    extensions: &'native mut NativeContextExtensions<'native_lifetimes>,
+}
+
+impl RunContext<'_, '_, '_> {
+    // TODO: The Run Context should hold this, not go get it from the Loader.
+    fn vm_config(&self) -> &VMConfig {
+        self.loader.vm_config()
+    }
+}
+
+macro_rules! set_err_info {
+    ($frame:expr, $e:expr) => {{
+        let function = $frame.function();
+        $e.at_code_offset(function.index(), $frame.pc)
+            .finish($frame.location())
+    }};
+}
+
+/// Main loop for the execution of a function.
+///
+/// This runs a newly-made Machine until it is complete. It expects the Machine to have a current
+/// call frame set up, with no operands on the stack and no existing call stack. It runs in a loop,
+/// calling `step`, until the call stack is empty.
+pub fn run(
+    start_state: MachineState,
+    data_store: &impl DataStore,
+    gas_meter: &mut impl GasMeter,
+    extensions: &mut NativeContextExtensions,
+    loader: &Loader,
+) -> VMResult<Vec<Value>> {
+    let link_context = data_store.link_context();
+    let resolver = start_state.current_frame.resolver(link_context, loader);
+
+    let mut run_context = RunContext {
+        loader,
+        link_context,
+        defn_resolver: resolver,
+        extensions,
+    };
+
+    let mut state = start_state;
+
+    if DEBUG_STEP_PRINT {
+        let mut buf = String::new();
+        let _ = state.debug_print_stack_trace(&mut buf, loader);
+        println!("Call Frame:\n{:?}", state.current_frame);
+        println!("{buf}");
+        buf.clear();
+    }
+
+    // Run until we're done or we produce an error and bail.
+    while step(&mut state, &mut run_context, gas_meter)? != StepStatus::Done {
+        if DEBUG_STEP_PRINT {
+            let mut buf = String::new();
+            println!("-------------------------------------");
+            println!("Call Frame:\n{:?}", state.current_frame);
+            let _ = state.debug_print_stack_trace(&mut buf, loader);
+            println!("{buf}");
+            buf.clear();
+        }
+        continue;
+    }
+
+    // When we are done, grab the operand stack as the return type.
+    let MachineState { operand_stack, .. } = state;
+    let RunContext { .. } = run_context;
+    Ok(operand_stack.value)
+}
+
+fn step(
+    state: &mut MachineState,
+    run_context: &mut RunContext,
+    gas_meter: &mut impl GasMeter,
+) -> VMResult<StepStatus> {
+    let fun_ref = state.current_frame.function();
+    let instructions = fun_ref.code();
+    let pc = state.current_frame.pc as usize;
+    assert!(pc <= instructions.len());
```

Comment:
worth to add a message? I am thinking mostly for an easier way to identify the problem in the log for any random person?